Chiheb feature create silver table date dimensions

ChihebDean opened the pull request.

ChihebDean committed:
"fix bronze data models"

ChihebDean committed:
"create data dimensions table"

saifeddine1992 commented on src/main/scala/com/fluidcode/processing/silver/CreateDataDimensionsTable.scala:
```diff
@@ -0,0 +1,28 @@
+import org.apache.spark.sql.{DataFrame, SparkSession}
+
+object CreateDataDimensionsTable {
+  def getDataDimensions(bronzeData: DataFrame): DataFrame = {
```
what is this "DataDimension" ???

saifeddine1992 commented on src/main/scala/com/fluidcode/processing/silver/CreateDataDimensionsTable.scala:
```diff
@@ -0,0 +1,28 @@
+    )
+    dataDimensions
+  }
+  def createDataDimensionsTable(conf: Configuration, spark: SparkSession): Unit = {
```
we don't have a DataDimension table

ChihebDean committed:
"edit silver data models"

ChihebDean committed:
"rename silver table from dataDimensions to dateDimension"

ChihebDean committed:
"fix object CreateDateDimensions, and test CreateDateDimensionsSpec"

ChihebDean committed:
"correct var name in function createDateDimensionsTable, and fix test description"

saifeddine1992 commented:
date dimension is wrong; check issue description.

ChihebDean committed:
"fix DateDimensions data models"

ChihebDean committed:
"-fix CreateDateDimensionsTable object:
-create function getDateDimensions to generate a date dimension table with date details for a given period.
-create function CalculateEasterDate, then use it to create User Defined Functions to check either it's a holiday or not.
-create function createDateDimensionsTable to fill silver table DateDimensionsTable"

saifeddine1992 commented on src/main/scala/com/fluidcode/processing/silver/CreateDateDimensionsTable.scala:
```diff
@@ -1,38 +1,100 @@
+    })
+
+    val dateDimensionDF = datesDF
+      .withColumn("date",col("date").cast(StringType))
+      .withColumn("description", date_format(to_date($"date", "yyyy-MM-dd"), "EEEE, dd MMMM yyyy"))
+      .withColumn("dayOfMonth", dayofmonth(col("date")))
+      .withColumn("dayOfWeek", date_format(col("date"), "EEEE"))
+      .withColumn("month",  date_format(to_date($"date", "yyyy-MM-dd"), "MMMM"))
+      .withColumn("year", year(col("date")))
+      .withColumn("quarter", quarter(col("date")))
+      .withColumn("isWeekend",
+        when(col("dayOfWeek")==="Sunday" || col("dayOfWeek")==="Saturday", true)
+          .otherwise(false))
+      .withColumn("isHoliday",
```
too many withColum function; replace it with another method

saifeddine1992 commented on src/main/scala/com/fluidcode/processing/silver/CreateDateDimensionsTable.scala:
```diff
@@ -1,38 +1,100 @@
-    val dateDimensions = getDateDimensions(bronzeData)
 
-    dateDimensions.writeStream
+  def createDateDimensionsTable(startDate: String, periodInDays: Int, conf: Configuration, spark: SparkSession): Unit = {
```
replace argument periodInDays with endDate

saifeddine1992 commented on src/main/scala/com/fluidcode/processing/silver/CreateDateDimensionsTable.scala:
```diff
@@ -1,38 +1,100 @@
+
+    val datesDF = datesSeq.toDF("date")
+
+    val isEasterDayUDF = udf((dateString: String) => {
```
annotation doesn't have the type of the output. and the function is not tested

saifeddine1992 commented on src/main/scala/com/fluidcode/processing/silver/CreateDateDimensionsTable.scala:
```diff
@@ -1,38 +1,100 @@
+      date.isEqual(easter)
+    })
+
+    val isPentecostDayUDF = udf((dateString: String) => {
+      val formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd")
+      val date = LocalDate.parse(dateString, formatter)
+      val year = date.getYear
+      val easter = calculateEasterDate(year)
+      date.isEqual(easter.plusDays(50))
+    })
```
function not tested

saifeddine1992 commented on src/main/scala/com/fluidcode/processing/silver/CreateDateDimensionsTable.scala:
```diff
@@ -1,38 +1,100 @@
+      date.isEqual(easter.plusDays(50))
+    })
+
+    val isAscensionDayUDF = udf((dateString: String) => {
+      val formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd")
+      val date = LocalDate.parse(dateString, formatter)
+      val year = date.getYear
+      val easter = calculateEasterDate(year)
+      date.isEqual(easter.plusDays(40))
+    })
```
same

saifeddine1992 commented:
put holiday names in French
